File: core/aarogya_core/types/experiment.py
"""Experiment record types — immutable once written."""

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

from aarogya_core.types.common import SchemaVersionMixin
from aarogya_core.types.ids import DatasetId, ExperimentId, ModelId, PipelineId


class ConfigRef(BaseModel):
    """Reference to a config file under configs/."""

    path: str
    hash: str | None = None


class ExperimentArtifactLayout(BaseModel):
    """Frozen on-disk layout under registry/experiments/EXP-#####/."""

    root: str
    config_dir: str = "config"
    results_dir: str = "results"
    report_dir: str = "report"
    artifacts_dir: str = "artifacts"


class Experiment(SchemaVersionMixin):
    """
    Tracked experiment metadata (registry record).

    Experiments are immutable after creation. Reruns allocate a new EXP-#####.
    """

    experiment_id: ExperimentId
    name: str
    status: Literal["planned", "running", "completed", "failed", "abandoned"] = (
        "planned"
    )
    hypothesis: str | None = None
    pipeline_id: PipelineId | None = None
    engine_id: str | None = None
    model_id: ModelId | None = None
    dataset_ids: list[DatasetId] = Field(default_factory=list)
    dataset_version: str | None = Field(
        default=None,
        description="Dataset version label used for this run, e.g. v3",
    )
    config: ConfigRef | None = None
    metrics: dict[str, float] = Field(default_factory=dict)
    artifact_path: str | None = None
    notes_path: str | None = None
    immutable: Literal[True] = True
    supersedes: ExperimentId | None = Field(
        default=None,
        description="Prior experiment this run replaces (never mutates the prior)",
    )
    forked_from: ExperimentId | None = None
    artifact_layout: ExperimentArtifactLayout | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None
    extras: dict[str, Any] = Field(default_factory=dict)

File: shared/types/experiment.ts
import { z } from "zod";

import { SCHEMA_VERSION } from "../constants/api";

export const experimentIdSchema = z.string().regex(/^EXP-\d{5}$/);
export const datasetIdSchema = z.string().regex(/^DATASET-\d{5}$/);
export const modelIdSchema = z.string().regex(/^MODEL-\d{5}$/);
export const assetIdSchema = z.string().regex(/^ASSET-\d{5}$/);
export const pipelineIdSchema = z.string().regex(/^PIPELINE-\d{5}$/);
export const versionIdSchema = z.string().regex(/^VERSION-\d{5}$/);
export const sampleIdSchema = z.string().regex(/^SAMPLE-\d{5,}$/);

export const experimentSchema = z.object({
  schema_version: z.literal(SCHEMA_VERSION).optional().default(SCHEMA_VERSION),
  experiment_id: experimentIdSchema,
  name: z.string(),
  status: z.enum(["planned", "running", "completed", "failed", "abandoned"]).default("planned"),
  hypothesis: z.string().nullable().optional(),
  pipeline_id: pipelineIdSchema.nullable().optional(),
  engine_id: z.string().nullable().optional(),
  model_id: modelIdSchema.nullable().optional(),
  dataset_ids: z.array(datasetIdSchema).default([]),
  dataset_version: z.string().nullable().optional(),
  metrics: z.record(z.number()).default({}),
  immutable: z.literal(true).default(true),
  supersedes: experimentIdSchema.nullable().optional(),
  artifact_path: z.string().nullable().optional(),
  notes_path: z.string().nullable().optional(),
});

export type Experiment = z.infer<typeof experimentSchema>;
export type ExperimentId = z.infer<typeof experimentIdSchema>;
export type DatasetId = z.infer<typeof datasetIdSchema>;
export type ModelId = z.infer<typeof modelIdSchema>;
export type AssetId = z.infer<typeof assetIdSchema>;
export type PipelineId = z.infer<typeof pipelineIdSchema>;
export type VersionId = z.infer<typeof versionIdSchema>;
export type SampleId = z.infer<typeof sampleIdSchema>;
